### ISSUE TITLE
[PATCH].callRemote() on pb.Root object, w/ kwarg 'name' raises type error

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11284]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| slyphon reported|
|-|-|
|Trac ID|trac#347|
|Type|defect|
|Created|2003-10-17 09:55:26Z|

Attachments:

* [pb.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1b0/1b04f217730b9e677ec7021775bb6c2c8f8ca543/b1a8f5246d48743dd29f330f32a7bd27d2a5a246.diff) (996 bytes) - added by slyphon on 2003-10-17 12:19:23Z - 
* [name_conflict_fix.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1b0/1b04f217730b9e677ec7021775bb6c2c8f8ca543/bdf285fcd002302e23ad7cd4c1b3c47c5ab8b46b.diff) (6301 bytes) - added by slyphon on 2003-10-18 09:14:50Z - 
* [name_conflict_fix.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1b0/1b04f217730b9e677ec7021775bb6c2c8f8ca543/740181e31236ec902411d0d63fc8b4a1ef1b6e53.diff) (6301 bytes) - added by slyphon on 2003-10-18 09:18:44Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__347 347
type__defect defect
reporter__slyphon slyphon
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1066384526000000 1066384526000000
changetime__1468840764490472 1468840764490472
version__ 
owner__ 
cc__spiv cc__itamarst cc__warner cc__slyphon
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
to make this bug happen:

save the following to make_bug_sender.py
---------------------------------------------
#!/usr/bin/env python2.3
# File: make_bug_sender.py
from twisted.spread import pb
from twisted.internet import reactor

class Sender(pb.Root):
    def remote_get_Foo(self, **kwargs):
        print 'get_Bookmarks', kwargs

def main():
    import pb_classes
    reactor.listenTCP(8789, pb.PBServerFactory(Sender()))
    reactor.run()

if __name__ == '__main__':
    main()

==================================================
save the following to make_bug_receiver.py

--------------------------------------------------
#!/usr/bin/env python2.3
# File: make_bug_receiver.py
from twisted.spread import pb
from twisted.internet import reactor

class Receiver:
    def __init__(self, factory):
        deferred = factory.getRootObject()
        deferred.addCallback(self.got_rootObj).addErrback(self.noObj)

    def noObj(self, reason):
        print 'noObject', reason
        reactor.stop()

    def got_rootObj(self, obj):
        print 'received root object'
        self.senderRoot = obj
        self.get_Foo(name='example')

    def got_remote_Foo(self, bm=None):
        print 'ran got_remote_Bookmarks'
        return bm

    def get_Foo(self, **kwargs):
        d = self.senderRoot.callRemote('get_Foo', **kwargs)
        d.addCallback(self.got_remote_Foo).addErrback(self.noObj)

if __name__ == '__main__':
    factory = pb.PBClientFactory()
    receiver = Receiver(factory)
    reactor.connectTCP('localhost', 8789, factory)
    reactor.run()
==========================================================

this is the traceback:

received root object
noObject [Failure instance: Traceback: exceptions.TypeError, callRemote() got
multiple values for keyword argument 'name'
/usr/lib/python2.3/site-packages/twisted/internet/defer.py:288:_runCallbacks
./make_bug_receiver.py:17:got_rootObj
./make_bug_receiver.py:24:get_Foo
]

---

I don't know if this is my fault or not, but itamar told me to file it as a bug.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
solved the problem by changing argument in RemoteReference.callRemote from name
to _name. seems to have fixed it.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
ran testall, caused no errors related to pb
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
the patch looks good to me. I suspect this is not a very common problem, but
when it *does* happen it will confuse people for longer than they need to be
confused.. and it's easy to fix.

We should probably look around for other places where we accept kwargs..
Deferred.addCallback comes to mind..
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
I've grepped around a bit.. here are the places where we might want
to privatize the "internal" arguments. In each one, we're offering
the user a relatively transparent way to cause a function to be
invoked elsewhere, and as it stands there are certain argument names
that they cannot use because of the extra parameters used by the setup
call.

 reactor.callFromThread(self, f, *args, **kw)
 reactor.addSystemEventTrigger(self, phase, eventType, f, *args, **kw)
 reactor.callWhenRunning(self, callable, *args, **kw)
 reactor.callLater(self, seconds, f, *args, **kw)
 reactor.callInThread(self, callable, *args, **kwargs)
 Deferred.addCallback(self, callback, *args, **kw)
  .addErrback
  .addBoth

In each case, the called method cannot use the other named parameters,
for instance a callLater() method cannot have a 'seconds' or 'f'
argument (at least it won't be accessible with a named parameter).

There are a few others I'm not sure about.. threadpool.py has a
dispatchWithCallback, xish/utility has a number of callbackish things,
and t.python.log has some callWithContext functions that may behave
the same way.

Of course we could always just say "don't do that", but I think the
question is how much time will be lost by people getting confused by
their otherwise-transparent method calls raising exceptions.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
I've committed your patch (along with a comment explaining the problem, and
a corresponding change to RemoteCacheObserver.callRemote. I'll leave the
bug open until we make a decision about those other methods.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
okay, i made the changes that warner suggested, i'm uploading a .diff, so you
guys can decide what you want to do (if you want to apply it, etc.)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
okay, i made the changes that warner suggested, i'm uploading a .diff, so you
guys can decide what you want to do (if you want to apply it, etc.)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
If we do decide to do something about this for all methods of this kind, we
should also update coding-standard.xhtml to mention it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Note that for deferred there's no need to patch -- after all, 
addCallback/addErrback/addBoth are just convenience methods, and .addCallbacks 
doesn't suffer from this problem.

Also, you possibly want to add 
twisted.application.internet.TimerService.__init__ to the list.
Again, in this case there's a workaround so I'm not exceedingly troubled
by this.

Perhaps the correct way is to allow functions to pass a dict explicitly?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Marking as patch, assigning to warner.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
ok, I patched the following methods:
 reactor.addSystemEventTrigger
 reactor.callWhenRunning
 reactor.callLater
 reactor.callInThread

I did not change Deferred.addCallback, because I don't think we want
to encourage code like:

 d.addCallback(self.getResult, 12, foo=blah, callback=other)

In these cases, the TypeError is raised immediately (instead of in
a Failure), so it should be pretty obvious to the user that they've
done something wrong.

I also added a section to howto/coding-standards to explain the issue,
but that's mostly for new code. There are some places (spread.refpath)
that don't follow these new standards, but I think most of them aren't
worth changing until someone reports a problem.

twisted/internet/base.py 1.67
doc/howto/coding-standard.xhtml 1.3
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|GitHub commented|
|-|-|

In changeset 4b2c6509d589adf065da772041946b1b26d4f85e
```
#!CommitTicketReference repository="" revision="4b2c6509d589adf065da772041946b1b26d4f85e"
Merge pull request #347 from twisted/8618-rodrigc-spread

Author: rodrigc
Reviewer: acabhishek942
Fixes: #8618

Fix MRO error on Python3 for class TestNode
```
